docs/apps/rbac.md

```diff
@@ -0,0 +1,538 @@
+
+To clarify exactly what checks are handled by what, see this table.
+
+ | Action                     | Permission class checks                                  | View/serializer checks                                                                     |
+ |----------------------------|----------------------------------------------------------|--------------------------------------------------------------------------------------------|
+ | POST / create              | Checking add permission for models with no parent object | Permission to the parent object (like organization)<br>Permission to other related objects |
+ | POST for special action    | Checks custom permission to obj                          | nothing                                                                                    |
+ | GET detail view            | Checks view permission to object                         | nothing                                                                                    |
+ | PUT / PATCH<br>Detail view | Checks edit permission to object                         | Permission to related objects                                                              |
+ | DELETE<br>Detail view      | Checks delete permission to object                       | nothing                                                                                    |
+
+Combining these functions, a hypothetical app setup might look like:
+
+```python
+from my_app.models import MyModel
+
+from rest_framework import viewsets
+from rest_framework import serializers
+
+from ansible_base.rbac.api.related import RelatedAccessMixin
+from ansible_base.rbac.api.permissions import AnsibleBaseObjectPermissions
+
+
+class MyModelSerializer(RelatedAccessMixin, serializers.ModelSerializer):  # (1)
+    class Meta:
+        model = MyModel
+        fields = '__all__'
+
+
+class MyModelViewSet(viewsets.ModelViewSet):
+    permission_classes = [AnsibleBaseObjectPermissions]  # (2)
+    serializer_class = MyModelSerializer
+
+    def get_queryset(self):
+        return MyModel.access_qs(self.request.user)  # (3)
+```
+
+This marks 3 different integration points.
+1. `RelatedAccessMixin` - checks access to related objects and gives access to users who create new objects
+2. `AnsibleBaseObjectPermissions` - checks object permission for detail views and global "add" permission
+3. `MyModel.access_qs` - filters the queryset to only objects the request user has permission to view
+
+### Creating a New Role Definition
+
+Developers should interact with `RoleDefinition` as an ordinary model.
+Giving and removing permissions should be handled by methods on a `RoleDefinition` instance.
+Ordinarily, the other models from this app should not be interacted with.
+
+NOTE: At various times a "role definition" may be referred to as just a "role".
+
+Roles are expected to be user-defined in many apps. Example of creating a custom role definition:
```

Comment:
Where should this process be done? In a migration, in apps.py, somewhere else?

Reply:
In a migration, as AWX & eda-server both migrate existing roles.